src/Lecture2.hs

```diff
@@ -74,7 +79,14 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt index xs
+  | index < 0          = (Nothing, xs)
+  | index >= numElems  = (Nothing, xs)
+  | otherwise          = (Just (xs !! index), xs')
+      where xs' = first ++ tail second
+            (first, second) = splitAt index xs
+            numElems = length $ take (index + 1) xs
```

Comment:
In the beginning I used `length xs` and this, of course, blew up with infinite lists: the next best solution was to do a `take` first.
Without using `length` I'm not sure how I can check if an index is off the list.
My non-functional side would search for something like a `try/catch` construct in Haskell, which I suspect is just not there?

Reply:
You don't need to check if the index is out of bounds separately. You can use `splitAt` and pattern-matching on the second element of the list. If the index is greater than the list length then `splitAt` returns the entire list as its first element of the pair and an empty list in the second place.

Otherwise, the head of the second list is exactly the element under the desired index.

`length` on lists requires full traversal of the list so there's no way to check if the index is out-of-bounds without traversing the list.

There exist a package that provides list that stores its length alongside the list but it's an external dependency:

* https://hackage.haskell.org/package/slist